Fix EZP-24701: Role versioning

glye opened the pull request:
Role version support is required in order to edit roles safely.
~~Will add support for role `is_new` and `version`, and use them as in legacy.~~
Implement versioning in a similar way as for ContentType, with a RoleDraft class.

This requires a schema change: `ALTER TABLE ezrole DROP PRIMARY KEY, ADD PRIMARY KEY(id,version)`
I added `data/updates/6.0/mysql.sql` containing the change and a description.

https://jira.ez.no/browse/EZP-24701

TODO
- [x] Fix 2 errors in existing tests
- [x] Add tests for drafts functionality
- [x] update script for Postgres
- [x] Fix REST tests
- [x] BC doc

andrerom commented on eZ/Publish/API/Repository/Values/User/RoleCreateStruct.php:
```diff
@@ -25,6 +25,20 @@
+     *
+     * @var bool
+     */
+    public $is_new = true;
```
isNew

glye replied:
@andrerom Right. Though in the case of user/usergroup for example, the api has no draft structs. Neither does Content and ContentTypeGroup, but ContentType does. Of these, only Content and ContentType support editing. Not clear to me yet why there are no ContentDrafts. Excavation continues.

gggeek replied:
I think that historically drafts might have been implemented for contenttypes because of the fact that the editing GUI was implemented via a multi-submit form (add a field, configure it, add another one, etc... then at the end click 'apply' and the existing contents get updated), whereas actions on other entities where applied atomically, with the whole entity being saved each time it is modified (there is no save-draft button when editing roles and policies). Iirc workflows also have a draft state, and they also have a multi-step editing process.

Since the idea for eZPlatform is not to be bound by
- legacy data structures
- any single GUI process
I think that the draft thing could be rethought (now or later) with a bit of a more consistent approach.

andrerom commented:
This is very much reflecting the db schema, should have a pass on this comparing it to concepts in other API classes instead to make it more consistent.

glye commented:
@andrerom OK, I see what you mean. I guess I can change the DB schema then, if it turns out that is necessary/beneficial. In legacy, both `is_new` and `version` are different from how versioning is usually done. (Though I hope we can get away with just ignoring is_new, and using version differently.)

andrerom commented:
@glye I didn't say you have to change db schema :) that is the benefit of layers, we can give things other names higher up to cleanup the approach and words used.

glye commented:
- ContentType(Draft) defines the version status in the SPI Type. STATUS_DEFINED: published, STATUS_DRAFT: temporary, STATUS_MODIFIED: deferred (not relevant for roles).
- Section doesn't do versioning.
- Content is implied to be draft until it is published. VersionInfo holds the states, draft/published/archived. Versioning is a lot more complex for Content, so the ContentType way seems a better model for Role.

@gggeek Consistency is good, as is automatic/implied draft saving (might want to do that for ContentType later). Note that legacy roles also have a draft state, though this is invisible to the user. It must have, since we don't want roles to apply until editing is complete.

andrerom commented:
> ContentType(Draft) defines the version status in the SPI Type. STATUS_DEFINED: published, STATUS_DRAFT: temporary, STATUS_MODIFIED: deferred (not relevant for roles).


> Section doesn't do versioning.

Any idea if @lolautruche will need that for the changes he is doing for UI? guess not as it is not multi part form.

glye commented:
Probably not, section holds only a name and that can be changed at any time without any direct consequences, and it's not multi-part.

gggeek commented:
Nitpick/OT: section holds a name AND identifier.
And, in my very personal and humble opinion, section identifier should be universally adopted in all APIs instead of section id.
Reasoning is simple: for any kind of data sharing across ezp instances, you always want to rely on user-assigned identifiers, not db-assigned ids.
Plus identifiers are much more understandable in code and config.

andrerom commented:
@gggeek  thats offtopic. But as mentioned to @glye on the [other PR](https://github.com/ezsystems/PlatformUIBundle/pull/294#issuecomment-131811038) we will as part of Role work try to fix the fact that identifier and name is stored as the same. Need translatable names and actual identifier field like found here: https://github.com/ezsystems/ezpublish-kernel/commits/backend_updates

andrerom commented:
Ok, here is alternative proposal for API changes (last two commits):
https://github.com/ezsystems/ezpublish-kernel/compare/proposed_api_changes

@glye @pspanja @gggeek review welcome, if it is an acceptable approach then @glye can cherry pick the commits relevant.

gggeek commented:
At first glance, proposed_api_changes seems nice  - I like the fact that groups can be skipped while loading by type, and the name+ts of creator and modifier (again for an OT comment: that's a piece of data we should have in all our non-content entities, to please security-paranoid customers)

andrerom commented:
> to please security-paranoid customers

OT: there are many ways to archive that, audit logging is also a tool here which is needed as we on these objects don't keep record of all prior versions, like we can on Content.

gggeek commented:
aha, I did not dare to mention that, but since you have been lured into my OT thread, yes: audit logging of changes to ANY and ALL entities is something I have recently been asked for. I am pretty sure that it is easier to do in eZ5 than it was in eZ4, short of storing the db query log, but I never tried to set it up. It might make for a good blog post...

glye commented:
@andrerom Thanks and +1 to role draft changes (apart from copypasta errata in last commit). As I read it this means doing it the same way as for ContentTypeDraft, which is reasonably straight forward.

glye commented:
Reverted first proposal, cherry-picked @andrerom's API change, and fixed the copypasta. (Now working on implementing the API changes.)

andrerom commented on eZ/Publish/API/Repository/RoleService.php:
```diff
@@ -26,8 +27,144 @@
+     * @since 6.0
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException if the authenticated user is not allowed to read this role
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException if a role with the given name was not found
```
name => id

andrerom commented on eZ/Publish/API/Repository/RoleService.php:
```diff
@@ -26,8 +27,144 @@
+     * @since 6.0
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException if the authenticated user is not allowed to update a role
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException if the name of the role already exists
```
name => identifier

andrerom commented:
@glye / @pspanja added/rebased commits from cba for name handling in [`proposed_api_changes`](https://github.com/ezsystems/ezpublish-kernel/compare/proposed_api_changes) branch, this implies we will aim to over time cherry pick in the other multiValued changes from cba's [`backend_updates`](https://github.com/ezsystems/ezpublish-kernel/compare/backend_updates) branch.

Looks ok?

glye commented:
@andrerom Looks OK from this distance, but there may be hidden devils I can't see at this detail level :)
E.g. there's no mention of how we should store/index/search these translation arrays. (Sure, it's a separate concern and a solvable one.)
As for `$mainLanguageCode` I hope we are not replicating all the bugs/confusion we had with this in legacy.

andrerom commented:
@glye This could be done as in content type, but you bring a important point, it will have impact on UI. So lets take versioning first and name/description work later.

glye commented:
Added RoleDraft implementation, added `status` property which is mapped to the legacy `version` DB field. Role draft creation in place.

How should we handle role edit conflicts and abandoned drafts? Suggestion: Show a conflict screen, which tells us who made the draft and when, and offer to either cancel editing, or continue (which will delete the other draft). To speed up delivery this can be a future improvement, and for now just continue editing.

andrerom commented on eZ/Publish/API/Repository/Values/User/Role.php:
```diff
@@ -38,6 +55,41 @@
+     *
+     * @var int One of Role::STATUS_DEFINED|Role::STATUS_DRAFT|Role::STATUS_MODIFIED
+     */
+    protected $status;
```
shouldn't this be on the RoleDraft object?
Role object implicit means it is published.

glye replied:
@andrerom Yes, but RoleDraft (as based on ContentTypeDraft) is just a wrapper around a Role object, and doesn't have any other properties. Isn't it good to be able to as a Role object that status it has? Particularly later if we are going to use STATUS_MODIFIED for e.g. deferred cache clearing?

andrerom replied:
ok, let's keep it like this.
Just a bit strange as you probably won't get published objects unless defined.

andrerom commented on eZ/Publish/Core/Repository/Values/User/RoleDraft.php:
```diff
@@ -0,0 +1,78 @@
+     */
+    public function getPolicies()
+    {
+        return $this->innerContentType->getPolicies();
```
?

glye commented:
Role draft creation works now. This introduces a DB change. Because of the drafts the `id` is no longer unique, so `version` is added to the primary key:
`ALTER TABLE ezrole DROP PRIMARY KEY, ADD PRIMARY KEY(id,version);`

Also it changes the Gateway implementation:
    -    public function loadRoleByIdentifier($identifier)
    +    public function loadRoleByIdentifier($identifier, $status = Role::STATUS_DEFINED)
to avoid some code duplication. Not sure if kosher.

andrerom commented:
> Because of the drafts the id is no longer unique, so version is added to the primary key:

how was this handled before?


> Also it changes the Gateway implementation:

should be fine on gateway level, but you probably need this for loading by `id` and not `identifier` right?

glye commented:
@andrerom In legacy, when making a temporary role version, the role ID was not copied, the draft got an autoincrement ID. Instead, `version` was set to be the role ID of the original role, that's how they were connected.

byId/byIdentifier: Possibly both. I needed it for byIdentifier to do role draft creation.

andrerom commented:
ok on the db change

> byId/byIdentifier: Possibly both. I needed it for byIdentifier to do role draft creation.

ok

glye commented:
Filled in many of the TODOs. Create and cancel (delete draft) works now.

glye commented:
Now publish works for realz.

glye commented:
Added dependencies:
https://github.com/ezsystems/PlatformUIBundle/pull/300
https://github.com/ezsystems/repository-forms/pull/41

glye commented:
Edit and publish already existing role works now (RoleService::createRoleDraftByRole etc).

glye commented:
Role edit and publish works now, versioning is implemented for policies so they are no longer wiped.

glye commented:
Fixed issue with publishing, due to wrong role caching.

glye commented:
Fixed bug that wipes policy limitations on editing. TODO: Fix the same for assignments.

glye commented:
Fixed bug that wipes role assignments on editing. TODO: Fix failing tests.

andrerom commented on eZ/Publish/Core/Repository/RoleService.php:
```diff
@@ -121,8 +126,386 @@ public function __construct(
+        }
+
+        try {
+            $existingRoleDraft = $this->loadRoleDraftByIdentifier($roleCreateStruct->identifier);
```
Seems to only check role drafts, while it also needs to check published roles, so maybe change the spi here to reflect that.

glye replied:
@andrerom Strictly speaking I should check both, but on creation there will never actually be any conflicts on non-draft roles. When creating new roles an md5-based internal identifier is used, and when creating a draft from an existing role the existing role identifier is known to be unique.

When updating the role however, I certainly need to check both, and react differently in each conflict case.

How would you change the SPI? How about just adding the non-draft check here in the RoleService?

andrerom replied:
The md5 stuff is in repository forms or platform ui controller code right? There is no knowledge about that here, and hence no assumptions on it should be made. The nature of layers, you only know what is going on below based on interface contracts, not above :)

glye replied:
s/ogres/Symfony stack/: http://bit.ly/1NEeyzJ

andrerom commented on eZ/Publish/Core/Repository/RoleService.php:
```diff
@@ -121,8 +126,386 @@ public function __construct(
+
+        if ($roleUpdateStruct->identifier !== null) {
+            try {
+                $existingRoleDraft = $this->loadRoleDraftByIdentifier($roleUpdateStruct->identifier);
```
same here

andrerom commented on eZ/Publish/Core/Repository/RoleService.php:
```diff
@@ -434,6 +825,32 @@ public function loadRole($id)
+     *
+     * @return \eZ\Publish\API\Repository\Values\User\RoleDraft
+     */
+    public function loadRoleDraftByIdentifier($identifier)
```
Could you mention a use case for adding this? I don't think it it needed besides possibility in spi to know if identifier is used and by which role, but might be wrong :)

glye replied:
@andrerom I guess it was added for completeness, and for detecting role edit conflicts.

andrerom replied:
> and for detecting role edit conflicts.

but where is this done now, is it done at several layers? cause here it is done on "Business Logic" [layer](https://doc.ez.no/pages/viewpage.action?pageId=11403666#eZPublish5Architecture-Introduction&Overview-Thetargetarchitecture:"eZPlatform"), aka the implementation of the API, so for that we only need to document exceptions to expect for this in create/update api, and have spi methods that allows us to check this (for drafts )

glye replied:
@andrerom Editing conflicts are not detected yet, there's a followup for that (not decided how we should deal with them UI-wise). Currently I'm just overwriting any existing drafts.

> it only checks against published items for identifier and remote id conflicts

I don't see how that can work here - the DB will refuse to store on draft conflicts, and we need to store drafts when editing...?

So, like this then?
- remove loadRoleDraftByIdentifier() from the RoleService, instead call SPI method directly
- on conflict, just throw exceptions in RoleService, and catch them in PlatformUI

andrerom replied:
> So, like this then? ..

yes something like that

> I don't see how that can work here - the DB will refuse to store on draft conflicts, and we need to store drafts when editing...?

And what is the problem? this is exactly what you are already doing in https://github.com/ezsystems/ezpublish-kernel/pull/1376/files#r38186539 or?

glye replied:
@andrerom You mentioned "only checks against published items", the PR is checking drafts. The problem, I think, in not specifically checking for each expected error case and just throwing exception on any DB error, is that the error message the user gets may not be specific enough to be useful.

glye commented:
The tests fail due to a DB issue, which is expected since I modified the primary key of the ezrole table, but the actual error makes no sense to me: "General error: 20 datatype mismatch" on an insert query which in my local tests works fine both before and after the primary key change.

glye commented:
Improved checking of existing roles by identifier, on create/update.

lolautruche commented on eZ/Publish/API/Repository/Values/User/Role.php:
```diff
@@ -23,7 +23,24 @@
+     *
+     * Reserved for future use, e.g. to delay cache invalidation after a role change.
+     */
+    const STATUS_MODIFIED = 2;
```
If we don't use it, then why having a constant for it?

glye replied:
@lolautruche I discussed with AR to use it for delayed/async role cache clearing. But we can of course remove it until it's needed.

glye commented:
@andrerom I need to get the DB schema change done on Travis. And how/where should I commit the change? Is there a way to commit the change so that Travis updates the schema by itself?

`ALTER TABLE ezrole DROP PRIMARY KEY, ADD PRIMARY KEY(id,version);`

lolautruche commented:
> Any idea if @lolautruche will need that for the changes he is doing for UI? guess not as it is not multi part form.

I don't see how it could be useful.

andrerom commented:
> I don't see how it could be useful.

Ok, me neither

> there a way to commit the change so that Travis updates the schema by itself?

See Pedro's password PR, as for humans we need to define where to place updates in 6.0, maybe data/updates/<mysql|postgres> for now

glye commented:
Removed STATUS_MODIFIED.

glye commented:
Added schema changes and BC doc.

glye commented:
@andrerom There are two types of test failures left now, and I could use some help with both:

1. "eZ\Publish\Core\Persistence\Cache\Tests\UserHandlerTest::testCreateRole
Expectation failed for method name is equal to <string:getItem> when invoked 1 time(s).
Method was expected to be called 1 times, actually called 0 times."
This happens only on Travis, not locally. Does it mean that it's looking for the method and not finding it? This method does indeed not exist.

2. A bunch of Postgres-only errors coming from a LEFT JOIN with two ON-conditions, producing invalid SQL. Doesn't Postgres support this? (Haven't found any indication of that yet.) E.g. https://github.com/ezsystems/ezpublish-kernel/pull/1376/files#diff-fa27bd8811515fc4cf3158f0e9ebd0daR109

andrerom commented on doc/bc/changes-6.0.md:
```diff
@@ -158,6 +158,13 @@ Changes affecting version compatibility with former or future versions.
+  is which. You should make sure you don't have any role drafts in the database when upgrading. To find drafts, select
+  entries in the `ezrole` table where `version` is not zero. Before this change, the `version` of a draft equals the
+  `id` of the role. The safe way to get rid of stray role drafts is to edit the role in question, and then cancel
+  editing.
```
we don't write bc notes for db changes, that is what db update.sql files are for.

glye replied:
Ok, but I need to put a descriptive text somewhere, since role drafts will not survive the update. Can write it as a comment in an SQL file. So, `data/updates/6.0/mysql.sql` then?

andrerom commented:
> ""eZ\Publish\Core\Persistence\Cache\Tests\UserHandlerTest::testCreateRole"  .. getItem ..

getItem refers to stash api, so where does "string" come from?

glye commented:
Thanks! I've wrapped multiple conditions in `$query->expr->lAnd()`, works well in local unit tests, hoping it satisfies postgres on travis.

glye commented:
The postgres tests are better now, but what remains makes no sense:
https://travis-ci.org/ezsystems/ezpublish-kernel/jobs/78362577

glye commented:
Fixed the test failure in `eZ\Publish\Core\Persistence\Cache\Tests\UserHandlerTest::testCreateRole`. `Userhandler::createRole()` no longer caches the role, because it may be a draft. So the test should not expect the cache methods to be called. All tests should pass now.

glye commented:
Existing tests pass now. Work in progress: Adding tests of new functionality, and fixing bugs as they show up.

glye commented:
Added more tests and bugfixes. Created `data/updates/6.0/mysql.sql` with alter table statement. Moved descriptive text from `changes-6.0.md` to this file.

glye commented:
Review ping @andrerom @lolautruche @pspanja @bdunogier @dpobel

bdunogier commented on eZ/Publish/API/Repository/RoleService.php:
```diff
@@ -26,8 +27,144 @@
+     * @since 6.0
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException if the authenticated user is not allowed to create a role
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException if the name of the role already exists or if limitation of the
```
I think that you can wrap @throws this way:

```
@throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
        if the name of the role already exists or if limitation of the same type
        is repeated in the policy create struct or if limitation is not allowed on module/function
```

glye replied:
OK

bdunogier commented on eZ/Publish/API/Repository/Tests/RoleServiceTest.php:
```diff
@@ -613,6 +959,47 @@ public function testAddPolicyUpdatesRole()
+
+        $roleCreate = $roleService->newRoleCreateStruct('newRole');
+
+        // @todo uncomment when support for multilingual names and descriptions is added
```
Maybe we should include an issue # in the comment, so that we can easily find those ?

glye replied:
OK

lolautruche commented on eZ/Publish/API/Repository/Values/User/RoleCreateStruct.php:
```diff
@@ -18,6 +18,13 @@
+     *
+     * @var int
+     */
+    public $id;
```
Why would we need an ID for a create struct ?

bdunogier replied:
+1 with @lolautruche

glye replied:
@lolautruche @bdunogier It is used when creating a role draft from an existing published role:
https://github.com/ezsystems/ezpublish-kernel/pull/1376/files#diff-c4d78f93737fdf4d5fd48c19e33b65c7R220

lolautruche replied:
See my comment about `createRole()` / `createRoleDraft()`.
We should not have an ID in a create/update struct. Problem comes with how role / role draft creation is made with your current change.

bdunogier replied:
Well, in that case, why not call it existingRoleId, or similar ? The reason is good, the name: not that much :-)

Would that work ?

lolautruche replied:
It's introducing an inconsistency with other structs and other repository services. AFAICT, we don't have any other structs holding an existing ID.

bdunogier commented on eZ/Publish/Core/SignalSlot/RoleService.php:
```diff
@@ -66,8 +67,243 @@ public function __construct(RoleServiceInterface $service, SignalDispatcher $sig
+    {
+        $returnValue = $this->service->createRoleDraft($roleCreateStruct);
+        $this->signalDispatcher->emit(
+            //TODO: CreateRoleDraftSignal? Or skip it altogether?
```
Think use-cases. Would signals in this area commonly need to distinguish drafts ?
Maybe an argument to the signal would be sufficient ?

glye replied:
@bdunogier I have no idea. To my mind there's no need to broadcast that we are creating drafts, but I'm sure someone somewhere could find a use for it.

lolautruche commented on eZ/Publish/Core/Persistence/Cache/UserHandler.php:
```diff
@@ -203,6 +246,30 @@ public function deleteRole($roleId)
+        $return = $this->persistenceHandler->userHandler()->publishRoleDraft($roleId);
+
+        $this->cache->clear('user', 'role', $roleId);
+        $this->cache->clear('user', 'role', 'assignments');
```
Maybe we could warmup the cache here as well?

glye replied:
OK

bdunogier commented on eZ/Publish/Core/SignalSlot/RoleService.php:
```diff
@@ -66,8 +67,243 @@ public function __construct(RoleServiceInterface $service, SignalDispatcher $sig
+    {
+        $returnValue = $this->service->updateRoleDraft($roleDraft, $roleUpdateStruct);
+        $this->signalDispatcher->emit(
+            //TODO: UpdateRoleDraftSignal? Or skip it altogether?
```
Can you actually update a Role that ain't a draft ? It would not make much sense, would it ?

glye replied:
@bdunogier Yes, until the deprecated methods for that are removed.

bdunogier commented:
Maybe the signals & draft part needs to be clarified. I'll be +1 once that is done.

lolautruche commented on eZ/Publish/API/Repository/RoleService.php:
```diff
@@ -26,8 +27,144 @@
+     *
+     * @return \eZ\Publish\API\Repository\Values\User\RoleDraft
+     */
+    public function createRoleDraft(RoleCreateStruct $roleCreateStruct);
```
If we want to be consistent with ContentTypes, this method should receive an existing `Role`, not a struct.
To create a brand new role draft from scratch, we should then have `createRole(RoleCreateStruct $roleCreateStruct)` which would return an empty `RoleDraft` object.

See [ContentTypeService::createContentTypeDraft()](https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/Repository/ContentTypeService.php#L1262) / [ContentTypeService::createContentType()](https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/Repository/ContentTypeService.php#L695)

glye replied:
@lolautruche The API changes where designed by @andrerom, and I think his reason for not changing `createRole()` was BC. Awaiting his feedback before I change anything here.

glye replied:
Thanks for all contributions - the conclusion is to align with ContentType where possible.

lolautruche commented on eZ/Publish/Core/Persistence/Legacy/User/Handler.php:
```diff
@@ -279,6 +355,38 @@ public function deleteRole($roleId)
+    {
+        $roleDraft = $this->loadRoleDraft($roleId);
+
+        $this->roleGateway->publishRoleDraft($roleDraft->id);
```
What about publishing a draft from an existing role? This will probably fail with a DB exception as the gateway just updates the role record, so then we end up with 2 records with same PKs.
See [how it is done in `Content\Type\Handler`](https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/Persistence/Legacy/Content/Type/Handler.php#L569-L582). We first delete the old one and then publish the new one.

glye replied:
OK (it works fine in testing, new and existing roles)

glye replied:
@lolautruche Actually, I'm doing the delete-and-publish, which is why it works. Only I'm doing it in the RoleService, not in the Handler. If it's the wrong place I can move it.

glye commented:
Warmup cache when publishing, reformat throw doc, add issue ref for multilingual followup.

glye commented:
Moved delete-existing-and-publish-draft logic from RoleService to the Handler.

glye commented:
Removed role ID from create struct, and related changes.
Consolidated *Role/*RoleDraft methods to one, adding status as parameter.

glye commented:
The only remaining test failures now are due to Travis failing to install jarnaiz/behat-junit-formatter.

glye commented:
Updated RoleService signals, made new ones as needed.

lolautruche commented:
What about schema + update script for Postgres?
Technically we should also provide the same for SQLite since our integration tests rely partly on it

glye commented:
Yep, need that, and BC doc, and REST test fixes.

lolautruche commented:
+1 with those TODOs fixed

pspanja commented on eZ/Publish/Core/Persistence/Legacy/User/Handler.php:
```diff
@@ -166,35 +167,80 @@ public function delete($userId)
 
     /**
-     * Loads a specified role by $roleId.
+     * Loads a specified role (draft) by $roleId.
```
`Loads a role by $roleId and $status.`

Below:
`@throws \eZ\Publish\API\Repository\Exceptions\NotFoundException If role with given status does not exist`

You should also return that information in the exception, see similar: https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/Persistence/Legacy/Exception/TypeNotFound.php

pspanja commented on eZ/Publish/API/Repository/RoleService.php:
```diff
@@ -26,23 +27,144 @@
+    public function createRoleDraft(Role $role);
+
+    /**
+     * Loads a role for the given id.
```
Loads a RoleDraft...

pspanja commented on eZ/Publish/API/Repository/RoleService.php:
```diff
@@ -26,23 +27,144 @@
+     * @since 6.0
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException if the authenticated user is not allowed to read this role
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException if a role with the given id was not found
```
If role draft can't be found...

pspanja commented on eZ/Publish/API/Repository/RoleService.php:
```diff
@@ -26,23 +27,144 @@
+    public function updatePolicyByRoleDraft(RoleDraft $roleDraft, Policy $policy, PolicyUpdateStruct $policyUpdateStruct);
+
+    /**
+     * Deletes the given role.
```
...given Role draft.

pspanja commented on eZ/Publish/API/Repository/Values/User/RoleDraft.php:
```diff
@@ -0,0 +1,23 @@
+ * @property-read mixed $id the internal id of the role
+ * @property-read string $identifier the identifier of the role
+ *
+ * @property-read array $policies an array of the policies {@link \eZ\Publish\API\Repository\Values\User\Policy} of the role.
```
The same already exist in the extended class, it should not be needed to redefine them.

pspanja commented on eZ/Publish/API/Repository/Values/User/RoleCreateStruct.php:
```diff
@@ -25,6 +25,13 @@
+     *
+     * @var int One of Role::STATUS_DEFINED|Role::STATUS_DRAFT
+     */
+    public $status = Role::STATUS_DRAFT;
```
Why is this needed? AFAICS this makes possible to create defined Role through `RoleService::createRole(RoleCreateStruct $roleCreateStruct)`, which returns RoleDraft.

glye replied:
It's a BC artifact from before the refactoring, will remove it.

pspanja commented on eZ/Publish/API/Repository/Values/User/RoleUpdateStruct.php:
```diff
@@ -23,4 +23,11 @@ class RoleUpdateStruct extends ValueObject
+     *
+     * @var int One of Role::STATUS_DEFINED|Role::STATUS_DRAFT
+     */
+    public $status;
```
This should not be here as drafts are created through `RoleService::createRoleDraft(Role $role)`.
But maybe I'm missing something?

glye commented:
Added REST tests for draft functionality
Role -> RoleDraft and `$status` doc fixes
Removed `$status` from create/update structs, and related code changes
Removed incorrect property doc from RoleDraft
Added RoleNotFound exception, to be able to output role status

pspanja commented on eZ/Publish/Core/Persistence/Legacy/User/Handler.php:
```diff
@@ -166,38 +168,82 @@ public function delete($userId)
+        $role = $this->mapper->createRoleFromCreateStruct(
+            $createStruct
+        );
+        $role->id = $roleId;
```
What about creating a first draft, coming from `createRole(RoleCreateStruct $createStruct)`?
In this case an ID should be set, returned from the gateway.
You should also test this, see for example: https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/API/Repository/Tests/ContentTypeServiceTest.php#L123

pspanja replied:
Ok, I see now that is done in the gateway, ignore the ID part :)
But it should be tested anyway.

glye replied:
@pspanja Test what, exactly? (I've added a test for `createRoleDraft()`.)

pspanja replied:
Values of the created Role, see linked example for ContentTypeGroup.

glye replied:
Added test for this.

pspanja commented on eZ/Publish/Core/Persistence/Legacy/User/Handler.php:
```diff
@@ -253,29 +300,86 @@ public function loadRoles()
+     * @param int $status One of Role::STATUS_DEFINED|Role::STATUS_DRAFT
      */
-    public function deleteRole($roleId)
+    public function deleteRole($roleId, $status = Role::STATUS_DEFINED)
```
This should remove policies for the given version only, `$status` is not passed to the `Gateway::removePolicy()`.
The case should be tested as well.

glye replied:
Added test and fixed.

glye commented:
Fixed `deleteRole()` ignoring status, with test case.

pspanja commented on eZ/Publish/Core/Repository/RoleService.php:
```diff
@@ -157,23 +165,333 @@ public function createRole(APIRoleCreateStruct $roleCreateStruct)
+                 * - The ID of the two are NOT the same (otherwise it's not an editing conflict)
+                */
+                $existingRole = $this->loadRoleByIdentifier($roleUpdateStruct->identifier);
+                if ($existingRole instanceof APIRole && $existingRole->id != $loadedRoleDraft->id) {
```
It is not necessary to check for the `instanceof`, we can rely on `loadRoleByIdentifier()` returning a Role or throwing an exception.

pspanja commented on eZ/Publish/API/Repository/Values/User/Role.php:
```diff
@@ -38,6 +48,41 @@
+     *
+     * @var \DateTime
+     */
+    protected $creationDate;
```
This and `$modificationDate`, `$creatorId` and `$modifierId` below are not used and will not be used in 6.0 (we would need to change the DB for that), so I would rather remove them for the time being.

pspanja commented on eZ/Publish/Core/Repository/RoleService.php:
```diff
@@ -358,6 +680,8 @@ protected function internalDeletePolicy(APIPolicy $policy)
      * Updates the limitations of a policy. The module and function cannot be changed and
      * the limitations are replaced by the ones in $roleUpdateStruct.
      *
+     * @deprecated since 6.0, use {@link updatePolicyByRoleDraft()} instead.
```
Should probably not deprecate for the method that is not implemented yet :)

glye commented:
Added test for createstruct values vs role values, Removed unneccesary checks, removed unused role properties, un-deprecated `updatePolicy()` which should still be used.

glye commented:
Fixed the REST tests now, at least it works locally.

glye commented:
Added Postgres update script.

@lolautruche Does it make sense to have an update script for SQLite, if it's only used dynamically?
There are no schema files for Postgres and SQLite, there is only `data/mysql/schema.sql`. Should that be fixed?
There are fixture files (`eZ/Publish/Core/Persistence/Legacy/Tests/_fixtures/schema.*`) and I have already set the new primary key in all of these.

lolautruche commented:
> There are no schema files for Postgres and SQLite, there is only data/mysql/schema.sql. Should that be fixed?

I guess so, at least for Postgres.

As for SQLite, I'm not sure. What's your take @andrerom @pspanja @bdunogier ?

pspanja commented:
There is no need for SQLite upgrade script, we use it for tests only.

glye commented:
Okay, then TODO is to add a postgres schema file, but that belongs in a different PR I think.

glye commented:
Added BC doc.

andrerom commented:
+1, solid work @glye, and indeed the postgres schema is part of different epic :)

lolautruche commented:
Still +1. Bravo @glye !

pspanja commented:
+1 Nice work @glye :)

glye commented:
Damn, forgot to squash :/
